// @fiscora/api/src/modules/xml-audit/sat-matrix/complementos/cartaporte-matrix.rules.ts
import type { SatMatrixRule } from "../sat-matrix.types.js";

export const CARTA_PORTE_MATRIX_RULES: SatMatrixRule[] = [
  {
    id: "CP-001",
    source: "FISCORA_FORENSIC",
    officialCode: "CP-01",
    officialField: "TipoDeComprobante",
    officialDescription: "Carta Porte solo en Tipo I o T",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_TYPE_INVALID"],
    severityAlignment: "SAME",
    testCases: ["AC"],
  },
  {
    id: "CP-002",
    source: "FISCORA_FORENSIC",
    officialField: "CartaPorte.Version",
    officialDescription: "Versión faltante o desconocida",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_VERSION_MISSING", "CARTA_PORTE_VERSION_INVALID"],
    severityAlignment: "SAME",
  },
  {
    id: "CP-003",
    source: "FISCORA_FORENSIC",
    officialField: "CartaPorte.Traslado",
    officialDescription: "Carta Porte internacional sin país/vía",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_INTERNACIONAL_SIN_PAIS_VIA"],
    severityAlignment: "SAME",
    testCases: ["FF"],
  },
  {
    id: "CP-004",
    source: "FISCORA_FORENSIC",
    officialField: "CartaPorte.TotalDistRec",
    officialDescription: "TotalDistRec faltante o no positivo",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_TOTALDISTREC_MISSING", "CARTA_PORTE_TOTALDISTREC_NON_POSITIVE"],
    severityAlignment: "SAME",
  },
  {
    id: "CP-005",
    source: "FISCORA_FORENSIC",
    officialField: "Ubicacion",
    officialDescription: "Ubicación sin TipoUbicacion",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_UBICACION_TIPO_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "CP-006",
    source: "FISCORA_FORENSIC",
    officialField: "Ubicacion.RfcRemitenteDestinatario",
    officialDescription: "Ubicación sin RFC/nombre cuando aplica",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_UBICACION_RFC_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "CP-007",
    source: "FISCORA_FORENSIC",
    officialField: "Ubicacion.FechaHoraSalidaLlegada",
    officialDescription: "Ubicación sin FechaHoraSalidaLlegada",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_UBICACION_FECHA_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "CP-008",
    source: "FISCORA_FORENSIC",
    officialDescription: "Destino sin distancia",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_DESTINO_SIN_DISTANCIA"],
    severityAlignment: "SAME",
    testCases: ["EX"],
  },
  {
    id: "CP-009",
    source: "FISCORA_FORENSIC",
    officialField: "CartaPorte.TotalDistRec",
    officialDescription: "TotalDistRec vs suma distancias mismatch",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_TOTALDISTREC_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["EY"],
  },
  {
    id: "CP-010",
    source: "FISCORA_FORENSIC",
    officialField: "CartaPorte.NumTotalMercancias",
    officialDescription: "NumTotalMercancias faltante",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_NUMMERCANCIAS_MISSING"],
    severityAlignment: "SAME",
    testCases: ["EZ"],
  },
  {
    id: "CP-011",
    source: "FISCORA_FORENSIC",
    officialField: "CartaPorte.NumTotalMercancias",
    officialDescription: "NumTotalMercancias mismatch vs suma",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_NUMMERCANCIAS_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["EZ"],
  },
  {
    id: "CP-012",
    source: "FISCORA_FORENSIC",
    officialField: "CartaPorte.PesoBrutoTotal",
    officialDescription: "PesoBrutoTotal mismatch vs suma",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_PESOBRUTOTOTAL_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["FA"],
  },
  {
    id: "CP-013",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.BienesTransp",
    officialDescription: "Mercancía sin BienesTransp",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_MERCANCIA_BIENSTRANS_MISSING"],
    severityAlignment: "SAME",
    testCases: ["AE"],
  },
  {
    id: "CP-014",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.ClaveUnidad",
    officialDescription: "Mercancía sin ClaveUnidad",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_MERCANCIA_CLAVEUNIDAD_MISSING"],
    severityAlignment: "SAME",
    requiresCatalog: true,
    requiredCatalogs: ["CLAVE_UNIDAD"],
    testCases: ["FB"],
  },
  {
    id: "CP-015",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.Cantidad",
    officialDescription: "Mercancía sin cantidad no positiva",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_MERCANCIA_CANTIDAD_NON_POSITIVE"],
    severityAlignment: "SAME",
  },
  {
    id: "CP-016",
    source: "FISCORA_FORENSIC",
    officialDescription: "Material peligroso sin clave/embalaje",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_MATERIAL_PELIGROSO_SIN_CLAVE", "CARTA_PORTE_MATERIAL_PELIGROSO_SIN_EMBALAJE"],
    severityAlignment: "SAME",
    testCases: ["FC"],
  },
  {
    id: "CP-017",
    source: "FISCORA_FORENSIC",
    officialField: "Autotransporte.Permiso",
    officialDescription: "Autotransporte sin permiso",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_AUTOTRANSPORTE_PERMISO_MISSING"],
    severityAlignment: "SAME",
    testCases: ["FD"],
  },
  {
    id: "CP-018",
    source: "FISCORA_FORENSIC",
    officialField: "Autotransporte.Vehiculo",
    officialDescription: "Autotransporte sin vehículo",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_AUTOTRANSPORTE_VEHICULO_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "CP-019",
    source: "FISCORA_FORENSIC",
    officialField: "Autotransporte.Seguro",
    officialDescription: "Seguro RC faltante",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_AUTOTRANSPORTE_SEGURO_MISSING"],
    severityAlignment: "SAME",
    testCases: ["FD"],
  },
  {
    id: "CP-020",
    source: "FISCORA_FORENSIC",
    officialDescription: "Operador sin licencia",
    module: "CARTA_PORTE",
    complementKey: "CARTA_PORTE",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["CARTA_PORTE_OPERADOR_SIN_LICENCIA"],
    severityAlignment: "SAME",
    testCases: ["FE"],
  },
];
