// @fiscora/api/src/modules/xml-audit/sat-matrix/complementos/comercioexterior-matrix.rules.ts
import type { SatMatrixRule } from "../sat-matrix.types.js";

export const COMERCIO_EXTERIOR_MATRIX_RULES: SatMatrixRule[] = [
  {
    id: "CE-001",
    source: "FISCORA_FORENSIC",
    officialCode: "CE-01",
    officialField: "ComercioExterior",
    officialDescription: "Exportacion 02 requiere complemento Comercio Exterior",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["EXPORTACION_WITHOUT_COMERCIO_EXTERIOR"],
    severityAlignment: "SAME",
    testCases: ["DY"],
  },
  {
    id: "CE-002",
    source: "FISCORA_FORENSIC",
    officialField: "ComercioExterior.TipoOperacion",
    officialDescription: "TipoOperacion faltante o desconocido",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_TIPO_OPERACION_MISSING", "COMERCIO_EXTERIOR_TIPO_OPERACION_INVALID"],
    severityAlignment: "SAME",
    testCases: ["CC", "CD", "CE"],
  },
  {
    id: "CE-003",
    source: "FISCORA_FORENSIC",
    officialField: "ComercioExterior.Version",
    officialDescription: "Versión faltante o desconocida",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_VERSION_MISSING", "COMERCIO_EXTERIOR_VERSION_INVALID"],
    severityAlignment: "SAME",
    testCases: ["CG", "CD"],
  },
  {
    id: "CE-004",
    source: "FISCORA_FORENSIC",
    officialField: "ComercioExterior.TotalUSD",
    officialDescription: "TotalUSD faltante o no positivo",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_TOTALUSD_MISSING", "COMERCIO_EXTERIOR_TOTALUSD_NON_POSITIVE"],
    severityAlignment: "SAME",
    testCases: ["CF", "FP"],
  },
  {
    id: "CE-005",
    source: "FISCORA_FORENSIC",
    officialField: "ComercioExterior.TipoCambioUSD",
    officialDescription: "TipoCambioUSD requerido con moneda no XXX",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_TIPOCAMBIOUSD_MISSING"],
    severityAlignment: "SAME",
    testCases: ["FG"],
  },
  {
    id: "CE-006",
    source: "FISCORA_FORENSIC",
    officialField: "ComercioExterior.CertificadoOrigen",
    officialDescription: "CertificadoOrigen=1 sin NumCertificadoOrigen",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_CERTIFICADO_SIN_NUMERO"],
    severityAlignment: "SAME",
    testCases: ["FH"],
  },
  {
    id: "CE-007",
    source: "FISCORA_FORENSIC",
    officialDescription: "Complemento Vacío",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_EMPTY_COMPLEMENT", "COMERCIO_EXTERIOR_INCOMPLETE"],
    severityAlignment: "SAME",
    testCases: ["CH"],
  },
  {
    id: "CE-008",
    source: "FISCORA_FORENSIC",
    officialField: "Receptor",
    officialDescription: "Receptor sin ResidenciaFiscal/NumRegIdTrib",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_RECEPTOR_SIN_RESIDENCIA", "COMERCIO_EXTERIOR_RECEPTOR_SIN_NUMREG"],
    severityAlignment: "SAME",
    testCases: ["FI"],
  },
  {
    id: "CE-009",
    source: "FISCORA_FORENSIC",
    officialDescription: "Sin mercancías",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_SIN_MERCANCIAS"],
    severityAlignment: "SAME",
    testCases: ["FJ"],
  },
  {
    id: "CE-010",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.NoIdentificacion",
    officialDescription: "Mercancía sin NoIdentificacion",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_MERCANCIA_SIN_NOIDENTIFICACION"],
    severityAlignment: "SAME",
    testCases: ["FK"],
  },
  {
    id: "CE-011",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.FraccionArancelaria",
    officialDescription: "Mercancía sin fracción arancelaria",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_MERCANCIA_FRACCION_MISSING"],
    severityAlignment: "SAME",
    testCases: ["FL"],
  },
  {
    id: "CE-012",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.FraccionArancelaria",
    officialDescription: "Fracción arancelaria formato inválido",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_FRACCION_FORMATO_INVALIDO"],
    severityAlignment: "SAME",
    testCases: ["FM"],
  },
  {
    id: "CE-013",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.CantidadAduana",
    officialDescription: "Cantidad aduana no positiva",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_CANTIDAD_ADUANA_NON_POSITIVE"],
    severityAlignment: "SAME",
    testCases: ["FK"],
  },
  {
    id: "CE-014",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.UnidadAduana",
    officialDescription: "Unidad aduana faltante",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_UNIDAD_ADUANA_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "CE-015",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.ValorUnitarioAduana",
    officialDescription: "Valor unitario aduana no positivo",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_VALOR_UNITARIO_NON_POSITIVE"],
    severityAlignment: "SAME",
  },
  {
    id: "CE-016",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.ValorDolares",
    officialDescription: "ValorDolares no positivo",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_VALORDOLARES_NON_POSITIVE"],
    severityAlignment: "SAME",
    testCases: ["FM"],
  },
  {
    id: "CE-017",
    source: "FISCORA_FORENSIC",
    officialField: "ComercioExterior.TotalUSD",
    officialDescription: "TotalUSD vs suma mercancías mismatch",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_TOTALUSD_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["FN", "FP"],
  },
  {
    id: "CE-018",
    source: "FISCORA_FORENSIC",
    officialDescription: "Moneda USD sin TotalUSD matching",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_USD_SIN_TOTALUSD"],
    severityAlignment: "SAME",
    testCases: ["FP"],
  },
  {
    id: "CE-019",
    source: "FISCORA_FORENSIC",
    officialDescription: "Concepto sin mercancía asociada",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_CONCEPTO_SIN_MERCANCIA", "COMERCIO_EXTERIOR_CONCEPTO_MERCANCIA_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["FO"],
  },
  {
    id: "CE-020",
    source: "FISCORA_FORENSIC",
    officialField: "Mercancia.ValorDolares",
    officialDescription: "ValorDolares mismatch calculado",
    module: "COMERCIO_EXTERIOR",
    complementKey: "COMERCIO_EXTERIOR",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["COMERCIO_EXTERIOR_VALORDOLARES_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["FM"],
  },
];
