// @fiscora/api/src/modules/xml-audit/sat-matrix/complementos/nomina-matrix.rules.ts
import type { SatMatrixRule } from "../sat-matrix.types.js";

export const NOMINA12_MATRIX_RULES: SatMatrixRule[] = [
  {
    id: "NOMINA-001",
    source: "FISCORA_FORENSIC",
    officialCode: "NOM-01",
    officialField: "TipoDeComprobante",
    officialDescription: "Tipo N requiere complemento Nómina",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_MISSING_COMPLEMENT"],
    severityAlignment: "SAME",
    testCases: ["BG"],
  },
  {
    id: "NOMINA-002",
    source: "FISCORA_FORENSIC",
    officialField: "TipoDeComprobante",
    officialDescription: "Tipo distinto N con complemento Nómina",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_ON_NON_NOMINA_TYPE"],
    severityAlignment: "SAME",
    testCases: ["IJ"],
  },
  {
    id: "NOMINA-003",
    source: "FISCORA_FORENSIC",
    officialField: "Moneda",
    officialDescription: "Moneda distinta de MXN en documentos de nómina",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_MONEDA_NOT_MXN"],
    severityAlignment: "SAME",
  },
  {
    id: "NOMINA-004",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.FechaPago",
    officialDescription: "FechaPago faltante en Nómina",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_MISSING_FECHAPAGO"],
    severityAlignment: "SAME",
    testCases: ["EM"],
  },
  {
    id: "NOMINA-005",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.FechaInicialPago",
    officialDescription: "FechaInicialPago > FechaFinalPago",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_FECHAINI_MAYOR_FECHAFIN"],
    severityAlignment: "SAME",
    testCases: ["EN"],
  },
  {
    id: "NOMINA-006",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.NumDiasPagados",
    officialDescription: "NumDiasPagados faltante o no positivo",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_NUMDIASPAGADOS_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "NOMINA-007",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.Receptor",
    officialDescription: "Receptor sin TipoRegimen o PeriodicidadPago",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_RECEPTOR_SIN_REGIMEN", "NOMINA_RECEPTOR_SIN_PERIODICIDAD"],
    severityAlignment: "SAME",
    testCases: ["AH", "BO"],
  },
  {
    id: "NOMINA-008",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.Receptor.NumEmpleado",
    officialDescription: "Receptor sin número de empleado",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_RECEPTOR_SIN_NUMEMPLEADO"],
    severityAlignment: "SAME",
  },
  {
    id: "NOMINA-009",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.Percepciones",
    officialDescription: "Percepciones sin TotalGravado/TotalExento",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_PERCEPCIONES_TOTALGRAVADO_MISSING", "NOMINA_PERCEPCIONES_TOTALEXENTO_MISSING"],
    severityAlignment: "SAME",
    testCases: ["AI", "AJ"],
  },
  {
    id: "NOMINA-010",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.Percepciones.TotalGravado",
    officialDescription: "TotalGravado mismatch vs suma percepciones",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_PERCEPCIONES_TOTALGRAVADO_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["EP"],
  },
  {
    id: "NOMINA-011",
    source: "FISCORA_FORENSIC",
    officialField: "Percepcion.TipoPercepcion",
    officialDescription: "Percepción sin TipoPercepcion o Clave",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_PERCEPCION_TIPOPERCEPCION_MISSING", "NOMINA_PERCEPCION_CLAVE_MISSING"],
    severityAlignment: "SAME",
    testCases: ["EQ"],
  },
  {
    id: "NOMINA-012",
    source: "FISCORA_FORENSIC",
    officialField: "Percepcion.Concepto",
    officialDescription: "Percepción sin Concepto",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_PERCEPCION_CONCEPTO_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "NOMINA-013",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.Deducciones",
    officialDescription: "Deducciones sin TotalImpuestosRetenidos cuando aplica",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_DEDUCCIONES_TOTALIMPUESTOSRETENIDOS_MISSING"],
    severityAlignment: "SAME",
    testCases: ["ES"],
  },
  {
    id: "NOMINA-014",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.Deducciones.TotalOtrasDeducciones",
    officialDescription: "TotalOtrasDeducciones mismatch vs suma",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_DEDUCCIONES_TOTALOTRASDEDUCCIONES_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["ER"],
  },
  {
    id: "NOMINA-015",
    source: "FISCORA_FORENSIC",
    officialField: "Deduccion.TipoDeduccion",
    officialDescription: "Deducción sin TipoDeduccion",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_DEDUCCION_TIPODEDUCCION_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "NOMINA-016",
    source: "FISCORA_FORENSIC",
    officialField: "Nomina.OtrosPagos",
    officialDescription: "OtrosPagos sin TipoOtroPago",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_OTROPAGO_TIPOOPAGO_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "NOMINA-017",
    source: "FISCORA_FORENSIC",
    officialField: "SubsidioAlEmpleo",
    officialDescription: "SubsidioAlEmpleo con importe pero sin Tipo 002",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_SUBSIDIO_SIN_TIPO"],
    severityAlignment: "SAME",
    testCases: ["EU"],
  },
  {
    id: "NOMINA-018",
    source: "FISCORA_FORENSIC",
    officialField: "OtrosPagos.TotalOtrosPagos",
    officialDescription: "TotalOtrosPagos mismatch vs suma",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_OTROPAGOS_TOTAL_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["ET"],
  },
  {
    id: "NOMINA-019",
    source: "FISCORA_FORENSIC",
    officialDescription: "Total CFDI vs subtotal nómina mismatch",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_TOTAL_CFDI_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["EV"],
  },
  {
    id: "NOMINA-020",
    source: "FISCORA_FORENSIC",
    officialDescription: "Complemento Nómina vacío o incompleto",
    module: "NOMINA",
    complementKey: "NOMINA_12",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["NOMINA_EMPTY_COMPLEMENT", "NOMINA_INCOMPLETE"],
    severityAlignment: "SAME",
    testCases: ["AG", "AH"],
  },
];
