// @fiscora/api/src/modules/xml-audit/sat-matrix/complementos/retenciones-matrix.rules.ts
import type { SatMatrixRule } from "../sat-matrix.types.js";

export const RETENCIONES20_MATRIX_RULES: SatMatrixRule[] = [
  {
    id: "RET-001",
    source: "FISCORA_FORENSIC",
    officialCode: "RET-01",
    officialField: "Retenciones.Version",
    officialDescription: "Versión faltante o desconocida",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MISSING_VERSION", "RETENCIONES_VERSION_INVALID"],
    severityAlignment: "SAME",
    testCases: ["DC", "CD"],
  },
  {
    id: "RET-002",
    source: "FISCORA_FORENSIC",
    officialField: "Retenciones.FolioInt",
    officialDescription: "FolioInt faltante",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MISSING_FOLIOINT"],
    severityAlignment: "SAME",
    testCases: ["DB"],
  },
  {
    id: "RET-003",
    source: "FISCORA_FORENSIC",
    officialField: "Retenciones.FechaExp",
    officialDescription: "FechaExp faltante/inválida",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MISSING_FECHA_EXP", "RETENCIONES_FECHA_EXP_INVALID"],
    severityAlignment: "SAME",
    testCases: ["FQ", "DF"],
  },
  {
    id: "RET-004",
    source: "FISCORA_FORENSIC",
    officialField: "Retenciones.CveRetenc",
    officialDescription: "CveRetenc faltante/desconocido",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MISSING_CVE_RETENC", "RETENCIONES_CVE_RETENC_UNKNOWN"],
    severityAlignment: "SAME",
    requiresCatalog: true,
    requiredCatalogs: ["RETENCIONES_CVE"],
    testCases: ["GJ"],
  },
  {
    id: "RET-005",
    source: "FISCORA_FORENSIC",
    officialField: "Retenciones.LugarExpRetenc",
    officialDescription: "LugarExpRetenc faltante",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_LUGAR_EXP_FALTANTE"],
    severityAlignment: "SAME",
    testCases: ["FQ"],
  },
  {
    id: "RET-006",
    source: "FISCORA_FORENSIC",
    officialField: "Emisor.RfcE",
    officialDescription: "Emisor sin RFC",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_EMISOR_SIN_RFC"],
    severityAlignment: "SAME",
    testCases: ["FR"],
  },
  {
    id: "RET-007",
    source: "FISCORA_FORENSIC",
    officialField: "Receptor.Nacionalidad",
    officialDescription: "Receptor sin Nacionalidad",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_RECEPTOR_SIN_NACIONALIDAD"],
    severityAlignment: "SAME",
    testCases: ["FS"],
  },
  {
    id: "RET-008",
    source: "FISCORA_FORENSIC",
    officialField: "Periodo",
    officialDescription: "Periodo incompleto o sin ejercicio",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_PERIODO_INCOMPLETO"],
    severityAlignment: "SAME",
    testCases: ["FU"],
  },
  {
    id: "RET-009",
    source: "FISCORA_FORENSIC",
    officialField: "Periodo",
    officialDescription: "MesIni > MesFin",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MESINI_MAYOR_MESFIN"],
    severityAlignment: "SAME",
    testCases: ["DF"],
  },
  {
    id: "RET-010",
    source: "FISCORA_FORENSIC",
    officialField: "Totales.MontoTotOperacion",
    officialDescription: "MontoTotOperacion faltante/no positivo",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MONTO_OPERACION_MISSING"],
    severityAlignment: "SAME",
    testCases: ["FV"],
  },
  {
    id: "RET-011",
    source: "FISCORA_FORENSIC",
    officialField: "Totales.MontoTotGrav",
    officialDescription: "MontoTotGrav mismatch",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MONTO_GRAV_MISMATCH"],
    severityAlignment: "SAME",
  },
  {
    id: "RET-012",
    source: "FISCORA_FORENSIC",
    officialField: "Totales.MontoTotExent",
    officialDescription: "MontoTotExent mismatch",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MONTO_EXENT_MISMATCH"],
    severityAlignment: "SAME",
  },
  {
    id: "RET-013",
    source: "FISCORA_FORENSIC",
    officialField: "Totales.MontoTotRet",
    officialDescription: "MontoTotRet faltante/mismatch",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_TOTAL_RET_MISSING", "RETENCIONES_TOTAL_RET_MISMATCH"],
    severityAlignment: "SAME",
    testCases: ["DE", "FX"],
  },
  {
    id: "RET-014",
    source: "FISCORA_FORENSIC",
    officialField: "ImpRetenidos.BaseRet",
    officialDescription: "ImpRetenidos sin BaseRet",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_IMPRET_BASE_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "RET-015",
    source: "FISCORA_FORENSIC",
    officialField: "ImpRetenidos.ImpuestoRet",
    officialDescription: "ImpRetenidos sin ImpuestoRet/desconocido",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_IMPUESTO_MISSING", "RETENCIONES_IMPUESTO_UNKNOWN"],
    severityAlignment: "SAME",
    requiresCatalog: true,
    requiredCatalogs: ["IMPUESTOS_RETENCIONES"],
  },
  {
    id: "RET-016",
    source: "FISCORA_FORENSIC",
    officialField: "ImpRetenidos.MontoRet",
    officialDescription: "ImpRetenidos sin MontoRet",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MONTORET_MISSING"],
    severityAlignment: "SAME",
  },
  {
    id: "RET-017",
    source: "FISCORA_FORENSIC",
    officialField: "ImpRetenidos.TipoPagoRet",
    officialDescription: "TipoPagoRet faltante/desconocido",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_TIPOPAGO_MISSING", "RETENCIONES_TIPOPAGO_UNKNOWN"],
    severityAlignment: "SAME",
    requiresCatalog: true,
    requiredCatalogs: ["PAGOS_RETENCION"],
  },
  {
    id: "RET-018",
    source: "FISCORA_FORENSIC",
    officialDescription: " Complemento vacío",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_EMPTY_COMPLEMENT"],
    severityAlignment: "SAME",
    testCases: ["DB"],
  },
  {
    id: "RET-019",
    source: "FISCORA_FORENSIC",
    officialDescription: "Impuestos sin montos cuando MontoTotRet > 0",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_SIN_IMPUESTOS_CON_MONTOTOTRET"],
    severityAlignment: "SAME",
  },
  {
    id: "RET-020",
    source: "FISCORA_FORENSIC",
    officialDescription: "Sello/certificado incompleto",
    module: "RETENCIONES",
    complementKey: "RETENCIONES_20",
    coverage: "COVERED_DIRECT",
    fiscoraFindingCodes: ["RETENCIONES_MISSING_SELLO_OR_CERT_REVIEW"],
    severityAlignment: "SAME",
    testCases: ["DG"],
    requiresCryptoValidation: true,
  },
];
